persist/fs/commitlog/reader.go

```diff
@@ -56,34 +61,84 @@ type commitLogReader interface {
 	Close() error
 }
 
+type readResponse struct {
+	series     Series
+	datapoint  ts.Datapoint
+	unit       xtime.Unit
+	annotation ts.Annotation
+	resultErr  error
+}
+
+type decoderArg struct {
+	bytes checked.Bytes
+	err   error
+}
+
+type readerMetadata struct {
+	sync.RWMutex
+	decoder encoding.Decoder
+	lookup  map[uint64]Series
+	wgs     map[uint64]*sync.WaitGroup
+}
+
 type reader struct {
-	opts            Options
-	bytesPool       pool.CheckedBytesPool
-	chunkReader     *chunkReader
-	sizeBuffer      []byte
-	dataBuffer      []byte
-	logDecoder      encoding.Decoder
-	metadataDecoder encoding.Decoder
-	metadataLookup  map[uint64]Series
+	opts          Options
+	numConc       int
+	bytesPool     pool.CheckedBytesPool
+	chunkReader   *chunkReader
+	dataBuffer    []byte
+	logDecoder    encoding.Decoder
+	decoderBufs   []chan decoderArg
+	outBufs       []chan readResponse
+	cancelCtx     context.Context
+	cancelFunc    context.CancelFunc
+	shutdownChan  chan error
```

Comment:
nit: I think we use `Ch` instead of `Chan` everywhere else in the codebase

Reply:
Cool good to know, I'll change that